File: scala/src/main/scala/info/ljungqvist/sun/JulianDate.scala
package info.ljungqvist.sun

import java.util.Date

/**
  * Created on 24/09/15.
  *
  * @author Petter Ljungqvist (petter.ljungqvist@terdelle.com)
  */
class JulianDate(val dayNumber: Double) {

    import JulianDate._

    lazy val toDate = new Date(((dayNumber - JD1970.dayNumber) * MILLISECONDS_PER_DAY).toLong)

    lazy val abs = new JulianDate(Math.abs(dayNumber))

    lazy val j2000: Double = dayNumber - JD2000.dayNumber

    def +(days: Double) = new JulianDate(dayNumber + days)

    def -(days: Double) = new JulianDate(dayNumber - days)

    def -(jd: JulianDate): Double = dayNumber - jd.dayNumber

    def ==(jd: JulianDate): Boolean = equals(jd)

    def >(jd: JulianDate): Boolean = dayNumber > jd.dayNumber

    def <(jd: JulianDate): Boolean = ! >(jd) && ! ==(jd)

    def <=(jd: JulianDate): Boolean = ! >(jd)

    def hourPart: Double = dayNumber - dayNumber.asInstanceOf[Long]

    def canEqual(other: Any): Boolean = other.isInstanceOf[JulianDate]

    override def equals(other: Any): Boolean = other match {
        case that: JulianDate =>
            (that canEqual this) &&
                dayNumber == that.dayNumber
        case _ => false
    }

    override def hashCode(): Int = {
        val state = Seq(dayNumber)
        state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
    }
}

object JulianDate {

    def apply(dayNumber: Double): JulianDate = new JulianDate(dayNumber)

    def apply(date: Date): JulianDate = JD1970 + date.getTime.toDouble / MILLISECONDS_PER_DAY

    def JD(dayNumber: Double): JulianDate = JulianDate(dayNumber)
    def JD(date: Date): JulianDate = JulianDate(date)

    //implicit def cast(double: Double): JulianDate = JulianDate(double)

    //implicit def cast(julianDate: JulianDate): Double = julianDate.dayNumber

    val JD0 = JD(0d)
    val JD1970 = JD(2440587.5d)
    val JD2000 = JD(2451545.0d)
    val MILLISECONDS_PER_DAY: Double = 24d * 60d * 60d * 1000d

}

File: kotlin/src/main/kotlin/info/ljungqvist/sun/JulianDate.kt
package info.ljungqvist.sun

import java.util.*
import kotlin.math.floor

data class JulianDate(val dayNumber: Double) : Comparable<JulianDate> {

    override fun compareTo(other: JulianDate): Int = dayNumber.compareTo(other.dayNumber)

    fun toDate(): Date = Date(((dayNumber - Y1970.dayNumber) * MILLISECONDS_PER_DAY).toLong())

    val hourPart: Double
            get() = dayNumber - floor(dayNumber)

    val j2000: Double
        get() = dayNumber - Y2000.dayNumber

    operator fun plus(days: Double): JulianDate = jd(dayNumber + days)
    operator fun minus(days: Double): JulianDate = jd(dayNumber - days)
    operator fun minus(jd: JulianDate): Double = dayNumber - jd.dayNumber

    companion object {
        val Y0 = jd(0.0)
        val Y1970 = jd(2440587.5)
        val Y2000 = jd(2451545.0)
    }

}

private fun jd(dayNumber: Double) = JulianDate(dayNumber)
private const val MILLISECONDS_PER_DAY: Double = 24.0 * 60 * 60 * 1000

fun Date.toJulianDate() = JulianDate.Y1970 + this.time.toDouble() / MILLISECONDS_PER_DAY